internal/levee.go

```diff
@@ -39,75 +41,210 @@ func run(pass *analysis.Pass) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	// TODO: respect configuration scope
 
+	// TODO: respect configuration scope
 	sourcesMap := pass.ResultOf[source.Analyzer].(source.ResultType)
 
-	// Only examine functions that have sources
-	for fn, sources := range sourcesMap {
-		for _, b := range fn.Blocks {
-			if b == fn.Recover {
-				// TODO Handle calls to sinks in a recovery block.
-				continue // skipping Recover since it does not have instructions, rather a single block.
+	for fn := range sourcesMap {
+		analyzeFn(fn, conf, pass)
+	}
+
+	return nil, nil
+}
+
+func analyzeFn(fn *ssa.Function, conf *config.Config, pass *analysis.Pass) {
+	graph := newGraph()
+
+	for _, p := range fn.Params {
+		if conf.IsSource(utils.Dereference(p.Type())) {
+			graph.addSource(p.Name(), p)
+		}
+	}
+
+	// this is needed to capture variables referred to by closures
+	for _, p := range fn.FreeVars {
+		switch t := p.Type().(type) {
+		case *types.Pointer:
+			if s, ok := utils.Dereference(t).(*types.Named); ok && conf.IsSource(s) {
+				graph.addSource(p.Name(), p)
 			}
+		}
+	}
```

Comment:
This work was already done by the `source` analyzer, which should remain the source of truth for Source identification.  Use `sourcesMap[fn]` for a list of `*Source`s.  Export `Source.node` if you need it for your graph.

Reply:
I'm not sure I see the benefits of keeping the `source` analyzer as the source of truth for Source identification. One benefit of having separate analyzers is that we can test them separately and use `analysistest` to test Source detection specifically. However, I think Source detection could be tested more simply via unit tests. If I replace the Requires on the `source` analyzer with a Requires on the `buildssa` analyzer and iterate directly on `buildssa.ResultType`'s `srcFuncs` field, the tests still pass. Given that, we could transform the `source` analyzer into a `source` package that would provide utilities for Source identification (if needed; this would be on top of config). To me the biggest benefit with this approach would be simpler code. Not that having separate analyzers makes things particularly complex.

Are there benefits/limitations to each approach that I am missing?

Reply:
Avoiding duplicated effort was my primary concern.  Whether that comes from a `source` analyzer or `source` package functions is somewhat immaterial to me.

> However, I think Source detection could be tested more simply via unit tests.

So far, immutableT and I haven't managed to find a clean way to unit-test SSA steps that don't rely on `analysis.Analyzer`s and that aren't either brittle or heavily hard-coded.  If you see a good path forward that doesn't include us writing an entire test harness, I'm all for it.

> To me the biggest benefit with this approach would be simpler code. Not that having separate analyzers makes things particularly complex.

Is there an extra negative in here?  This sounds like you prefer distinct analyzers as a means to simpler code, with which I agree.  But that also seems to contradict the previous portion of the paragraph.

Reply:
> So far, immutableT and I haven't managed to find a clean way to unit-test SSA steps that don't rely on analysis.Analyzers and that aren't either brittle or heavily hard-coded. If you see a good path forward that doesn't include us writing an entire test harness, I'm all for it.

I believe I do have an approach that isn't too brittle. See #34. If we want to work with the SSA, I don't see a way to avoid writing **some** boilerplate. In my opinion, if we have functions that operate directly on SSA values/instructions, then it makes more sense to write a bit of code to get just the SSA than to write an analyzer and use analysistest.

> Is there an extra negative in here? This sounds like you prefer distinct analyzers as a means to simpler code, with which I agree. But that also seems to contradict the previous portion of the paragraph.

I believe I didn't express myself clearly. I think having separate analyzers can lead to simpler code, but in this case I don't see a clear case for having a separate analyzer just for source identification, so to me it actually makes the code more complex. Identifying sources seems to me a simple task that can be done while constructing/traversing the graph, so having a separate analyzer just to do that seems like unnecessary complexity.